optuna/samplers/cmaes/sampler.py

```diff
@@ -0,0 +1,351 @@
+            independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        )
+        self._n_startup_trials = n_startup_trials
+        self._warn_independent_sampling = warn_independent_sampling
+        self._logger = optuna.logging.get_logger(__name__)
+        self._cma_rng = np.random.RandomState(seed)
+
+    def infer_relative_search_space(
+            self, study: 'optuna.Study', trial: 'optuna.structs.FrozenTrial',
+    ) -> Dict[str, BaseDistribution]:
+
+        # Import here to avoid circular imports without breaking a backward compatibility.
+        from optuna.samplers import intersection_search_space
+
+        search_space = {}  # type: Dict[str, BaseDistribution]
+        for name, distribution in intersection_search_space(study).items():
+            if distribution.single():
+                # `cma` cannot handle distributions that contain just a single value, so we skip
+                # them. Note that the parameter values for such distributions are sampled in
+                # `Trial`.
+                continue
+
+            if not isinstance(
+                    distribution,
+                    (
+                        optuna.distributions.UniformDistribution,
+                        optuna.distributions.LogUniformDistribution,
+                        optuna.distributions.DiscreteUniformDistribution,
+                        optuna.distributions.IntUniformDistribution,
+                    ),
+            ):
+                # Categorical distribution is unsupported.
+                continue
+            search_space[name] = distribution
+
+        if len(search_space) < 2:
+            self._logger.info("`CmaEsSampler` only supports two or more dimensional continuous "
+                              "search space. `{}` is used instead of `CmaEsSampler`.".format(
+                                  self._independent_sampler.__class__.__name__))
+            self._warn_independent_sampling = False
+            return {}
+
+        return search_space
+
+    def sample_relative(
+            self,
+            study: 'optuna.Study',
+            trial: 'optuna.structs.FrozenTrial',
+            search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+
+        if len(search_space) == 0:
+            return {}
+
+        completed_trials = [
+            t
+            for t in study.get_trials(deepcopy=False)
+            if t.state == TrialState.COMPLETE
+        ]
+        if len(completed_trials) < self._n_startup_trials:
+            return {}
+
+        ordered_keys = [key for key in search_space]
+        ordered_keys.sort()
+
+        optimizer = self._restore_or_init_optimizer(
+            completed_trials, search_space, ordered_keys
+        )
+
+        solution_trials = [
+            t
+            for t in completed_trials
+            if optimizer.generation == t.system_attrs.get("cma:generation", -1)
+        ]
+        if len(solution_trials) >= optimizer.population_size:
```

Comment:
I'm worried that there might be somewhat large amount of trials that are evaluated but ignored by this condition when distributed or parallel optimization is executed.

For example, let's consider the case both the population size and parallelism are `5`.
The worst case is,
1. The first four workers of a generation started and completed evaluation of their trials (i.e., objective function).
2. Then, the fifth worker started his trial evaluation.
3. Before the fifth worker completed the evaluation, other four workers started to evaluate next trials (in the same generation).
4. After that, the fifth worker complete the evaluation, and CMA-ES's generation was shifted to the next one.
    - The results of other four worker will be discarded.

I think that it's not so uncommon scenario. If the evaluation of objective function is heavy, the overhead might not be able to be negligible.

Reply:
The verify the above scenario, I wrote a simple parallel optimization code as follows:
```python
import collections
import optuna
import random
import time

def objective(trial):
    x = trial.suggest_uniform('x', -100, 100)
    y = trial.suggest_int('y', -10, 0)
    time.sleep(random.random())
    return x**2 + y

if __name__ == '__main__':
    # When the number of hyperparameter is `2`, the population size is set to `6`.
    population_size = 6

    study = optuna.create_study(sampler=optuna.samplers.CmaEsSampler())
    study.optimize(objective, n_trials=population_size*20, n_jobs=population_size)

    populations = collections.defaultdict(int)
    for t in study.trials:
        populations[t.system_attrs.get("cma:generation",-1)] += 1

    print("Generation count: {}".format(len(populations)))
    print("Actual population size per generation:")
    for generation, size in sorted(list(populations.items())):
        print("  - {}: {}".format(generation, size))

    print("Rate of wasted trials: {}".format(1.0 - len(populations) * population_size / len(study.trials)))
```

The output is shown below:
```console
Generation count: 12
Actual population size per generation:
  - -1: 6
  - 0: 11
  - 1: 11
  - 2: 11
  - 3: 11
  - 4: 11
  - 5: 11
  - 6: 9
  - 7: 11
  - 8: 11
  - 9: 11
  - 10: 6
Rate of wasted trials: 0.4
```
This result seems to indicate if population and parallelism are the same size, about 40% of the results of evaluated trials would be discarded. But please let me know if I misunderstand something.

Reply:
MEMO: The maximum wasted trial rate is `(parallelism - 1) / (population_size + parallelism - 1)`

Reply:
Hmm... Yes, you are right. And I agree it's not so uncommon scenario.

Reply:
Thank you for checking. This seems to be a difficult problem to solve completely, so careful writing of the documentation (for users) and leave a TODO comment (for developers) may be a practical solution.

Reply:
[memo] @nmasahiro tells me that the approach of the following paper might be the solution of this problem.

* https://www.ini.rub.de/upload/file/1511451186_615a08ab0d795223958f/glasmachers2013b.pdf

Reply:
Interestring. Thank you for sharing the information.

Reply:
BTW, [kurobako](https://github.com/sile/kurobako) supports to benchmark algorithm in an concurrent and asynchronous execution (e.g., `$ kurobako studies --concurrency 10 --solvers ... --problems ...`).
I think that the feature might help you If you would implement the above asynchronous algorithm in the future (the behavior of `--concurrency` option hasn't been tested enough though...).

Reply:
Good suggestion. I executed kurobako benchmark with `--concurrency 10` option at https://github.com/c-bata/cmaes/pull/22. The benchmark results are below.

| benchmark function | [without concurrency optiion](https://github.com/c-bata/cmaes/pull/21) | [with --concurrency 10](https://github.com/c-bata/cmaes/pull/22) |
| --- | --- | --- |
| Six-Hemp camel | ![](https://camo.githubusercontent.com/b57709097739c7e2b70c08a64a34ef3f5b090bfc/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6b75726f62616b6f2d7265706f7274732f632d626174612f636d6165732f7369782d68656d702d63616d656c2d366539373937643865616638303863313761316439383738343835653833393162323038343136662e706e67) | ![](https://camo.githubusercontent.com/f319239a8a9e31e0c66bf97aceb061579ccf4007/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6b75726f62616b6f2d7265706f7274732f632d626174612f636d6165732f7369782d68656d702d63616d656c2d663066653864343536646366643562393034663836356239633131623533376333333433343064632e706e67) |
| Himmelblau | ![](https://camo.githubusercontent.com/d685d9ed8f0f78bb0acca41113b2d3ce5142a4d8/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6b75726f62616b6f2d7265706f7274732f632d626174612f636d6165732f68696d6d656c626c61752d366539373937643865616638303863313761316439383738343835653833393162323038343136662e706e67) | ![](https://camo.githubusercontent.com/73cf73b09f4de777e9f73a5798a1a5e7a8bbca2c/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6b75726f62616b6f2d7265706f7274732f632d626174612f636d6165732f68696d6d656c626c61752d663066653864343536646366643562393034663836356239633131623533376333333433343064632e706e67) |
| Rosenbrock | ![](https://camo.githubusercontent.com/9b55c99428bdcb292e75af286502fe6ad6ccc9ed/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6b75726f62616b6f2d7265706f7274732f632d626174612f636d6165732f726f73656e62726f636b2d366539373937643865616638303863313761316439383738343835653833393162323038343136662e706e67) | ![](https://camo.githubusercontent.com/aef77ddbe2bc8b680149c07e822da40f446ac1f4/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f6b75726f62616b6f2d7265706f7274732f632d626174612f636d6165732f726f73656e62726f636b2d663066653864343536646366643562393034663836356239633131623533376333333433343064632e706e67) |

> the behavior of --concurrency option hasn't been tested enough though...

When compared with previous benchmark results that don't specify `--concurrency` option like https://github.com/c-bata/cmaes/pull/21, I couldn't find a difference.

Reply:
Thank you for sharing the benchmark result. 

> I couldn't find a difference.

Hmm..., I will investigate it in the near future.

Reply:
Sorry for the delayed response.
I fixed a bug of `kurobako` relating to the `--concurrency` option and run the following benchmark:
```console
$ kurobako --version
kurobako 0.1.9

$ PROBLEM=$(kurobako problem sigopt himmel-blau)

// --concurrency=1
$ kurobako studies --solvers $(kurobako solver --name 'CMA-ES (concurrency=1)' command solver.py) --problems $PROBLEM --concurrency 1 --repeats 100 --budget 100 | kurobako run > cma-es-1.json

// --concurrency=10
$ kurobako studies --solvers $(kurobako solver --name 'CMA-ES (concurrency=10)' command solver.py) --problems $PROBLEM --concurrency 10 --repeats 100 --budget 100 | kurobako run > cma-es-10.json

// --concurrency=20
$ kurobako studies --solvers $(kurobako solver --name 'CMA-ES (concurrency=20)' command solver.py) --problems $PROBLEM --concurrency 20 --repeats 100 --budget 100 | kurobako run > cma-es-20.json

// plot
$ cat cma-es-*.json | kurobako plot curve --ymin 0
```

The following is the benchmark result plot. It seems that the concurrency level affected the optimization performance of CmaEsSampler.
![sigopt-evalset-himmelblau-dim-2-5b8dcbccb4b3ca005649ddded779aaac65823a3cfefb3740616632df14355741](https://user-images.githubusercontent.com/181413/77989737-78d99e00-735a-11ea-9a20-52f81eaa0039.png)

---

```python
#! /usr/bin/env python3
#
# filename: solver.py
#
# $ pip install git+https://github.com/sile/kurobako-py
# $ pip install git+https://github.com/c-bata/optuna@cmaes
from kurobako import solver
from kurobako.solver.optuna import OptunaSolverFactory
import optuna

optuna.logging.set_verbosity(optuna.logging.WARNING)


def create_study(seed):
    sampler = optuna.samplers.CmaEsSampler(seed=seed)
    return optuna.create_study(sampler=sampler)


if __name__ == "__main__":
    factory = OptunaSolverFactory(create_study, use_discrete_uniform=args.use_discrete_uniform)
    runner = solver.SolverRunner(factory)
    runner.run()
```